### ISSUE TITLE
Meterpreter vs Kex

### ISSUE THREAD

**Issue by KevinF4** [rapid7/metasploit-framework#15203]
what is the difference between a shell meterpreter exe that you run, and a meterpreter at the kernel level as in eternal blue? I know they are very rare to get, but what can one do that the other cannot?

**Comment by bcoles** [comment]
Meterpreter is an application with built-in functionality for post-exploitation tasks and support for command and control via a remote channel.

Shell payloads and Meterpreter payloads can be executed within process memory (ie, EternalBlue) rather than loaded from an executable file on disk (ie, `meterpreter.exe`).

For differences between userland (ring 3) and kernel (ring 0) privileges, refer to:

* https://unix.stackexchange.com/questions/137820/whats-the-difference-of-the-userland-vs-the-kernel

In short, the operating system kernel has control over the entire system: all memory (all processes) and all devices (network traffic, files on storage devices, ...); where as userland is restricted by the kernel. Users can interact with system memory and devices based on various access controls enforced by the kernel.

**Comment by KevinF4** [comment]
> Meterpreter is an application with built-in functionality for post-exploitation tasks and support for command and control via a remote channel.
> 
> Shell payloads and Meterpreter payloads can be executed within process memory (ie, EternalBlue) rather than loaded from an executable file on disk (ie, `meterpreter.exe`).
> 
> For differences between userland (ring 3) and kernel (ring 0) privileges, refer to:
> 
> * https://unix.stackexchange.com/questions/137820/whats-the-difference-of-the-userland-vs-the-kernel
> 
> In short, the operating system kernel has control over the entire system: all memory (all processes) and all devices (network traffic, files on storage devices, ...); where as userland is restricted by the kernel. Users can interact with system memory and devices based on various access controls enforced by the kernel.

meterpreter (exe) that we generate in the metasploit runs in the userland right? but the eternal blue when it runs successfully, runs in kernel mode, but what I keep thinking about is the amount of things that meterpreter can already do in userland, post modules, etc. can you give me some examples of what we could do with a kernel-mode meterpreter, that we can't do in userland?

**Comment by OJ** [comment]
When kernel exploits run, there's kernel shellcode that gets executed, this is not part of Meterpreter. This is something specific to kernel land and is usually minimal. The most common shellcode that's used simple changes the token of the Meterpreter process so that it runs at `NT AUTHORITY\SYSTEM`, effectively elevating privileges directly in kernel land. Sometimes more complex shellcode is needed, such as to create a process first. It varies depending on the exploit. Regardless, Meterpreter doesn't (and please, shouldn't!) ever run in kernel land.

**Comment by KevinF4** [comment]
> When kernel exploits run, there's kernel shellcode that gets executed, this is not part of Meterpreter. This is something specific to kernel land and is usually minimal. The most common shellcode that's used simple changes the token of the Meterpreter process so that it runs at `NT AUTHORITY\SYSTEM`, effectively elevating privileges directly in kernel land. Sometimes more complex shellcode is needed, such as to create a process first. It varies depending on the exploit. Regardless, Meterpreter doesn't (and please, shouldn't!) ever run in kernel land.

thanks for the explanation i'm still learning about it, i'm a student, but just a final question, why should meterpreter never run in kernel mode?

**Comment by bcoles** [comment]
Perhaps someone on IRC or Slack can help you out with your questions:

* https://www.metasploit.com/help

**Comment by timwr** [comment]
Fyi this is the kernel shellcode that eternalblue uses: https://github.com/rapid7/metasploit-framework/blob/master/external/source/shellcode/windows/multi_arch_kernel_queue_apc.asm
